### PR TITLE
bugfix/121-result-distance-ordering

### LINKED ISSUE #47: Order by relevance
e.g. a search for "age concern" should show those with both words next to each other first.

"needs more info"

### LINKED ISSUE #121: Results for ”Start typing an address” not accurate
Examples:
New Plymouth Budget Service located at 32 Leach Street New Plymouth.

Search 1 (distance 25km).

Search by entering (and selecting) 36 Leach Street New Plymouth

Expected result: New Plymouth Budget Service showing at top

Results: 1 result = Whanau Pakari with address as 118 Tukapa Street New Plymouth.

Search 2 (distance 25km).

Select Category = Budgeting/Financial Capability and enter (select) 36 Leach Street New Plymouth

Expected result: New Plymouth Budget Service showing at top

Results: 1 result = New Plymouth Budget Advisory Service

General issue

Search 3 (distance 100km).
Select Category = Child Care
Search by entering (and selecting) 56 the Terrace Wellington Central

Expected result: 115 results

Results: 4 results only

Search 4 (distance 100km) 
Select Category = Child Care
Search by entering (and selecting) Wellington Central

Expected result: 115 results

Results:  8 results only (no Lower Hutt based providers but 1 Upper Hutt based)

Search 5
Select Category = Child Care
Search by entering (and selecting) Lower Hutt

Expected result: 26 results

Results:  9 results only (no Lower Hutt based providers)
Note there are 26 “Local” providers under Child Care for Lower Hutt on the Family Services Directory (FSD).  The first result on the FSD is showing on Searchinator for a topic search (searched for Greenstone).  Checked the data.govt extract and it is in the Extract with Child Care as a Level 1.

### PR DESCRIPTION
**Pull request by lamorrison**
Updates the CKAN queries to work around a problem where the most relevant results are listed last (`sort by score asc`, basically). Each search now makes two requests - the first one with `limit=0` to determine how many results there are without actually fetching the results, and then one with `offset=${resultCount-limit}&limit=${limit}` to get the LAST page of results.

This also closes #47 which was phrase searching not working because of the same reverse ranking problem.

This also includes changes to search by location if no keywords are supplied.

closes #121 
closes #47

### CODE REVIEW

**Comment by Br3nda** [comment]
This pull request **fixes 1 alert** when merging aacb8b73ad6c8971bf777b153955724d1e9cab50 into c67e87d80d5f4cf3e1e2e68facc9c2b36327db06 - [view on LGTM.com](https://lgtm.com/projects/g/ServiceInnovationLab/FSD/rev/pr-2a68126471c24736b5f6648707b2c4db7ed46de3)

**fixed alerts:**

* 1 for Missing variable declaration

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: features/main-page.feature [hunk 1]
```diff
@@ -9,6 +9,4 @@ Feature: Stuff on the main page
   Scenario: Search controls
     Then I should see the keyword search box
     And I should see the location search box
-    And I should see some category selectors
-    And I should see the radius selector
```
Comment (lamorrison) [inline features/main-page.feature]: This reflects Jake's recent changes to radius display, not related to the point of this PR

**Comment by Br3nda** [comment]
This pull request **fixes 1 alert** when merging 225eadaa25829bfcd53db9b39e43fcad6818c1f3 into 534effc3d673fa340eabd5fbb37970133d5dca78 - [view on LGTM.com](https://lgtm.com/projects/g/ServiceInnovationLab/FSD/rev/pr-4f812d7f4f5b7b648ea218ce2e779a7bde44f719)

**fixed alerts:**

* 1 for Missing variable declaration

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: src/utilities/__tests__/url.test.js [hunk 2]
```diff
@@ -75,94 +75,39 @@ describe('url.js', () => {
       }).not.toThrow();
     });
 
-    it("doesn't clarify keyword when it's null", () => {
```
Comment (lamorrison) [inline src/utilities/__tests__/url.test.js]: This behaviour no longer applies

File: src/utilities/__tests__/url.test.js [hunk 3]
```diff
@@ -75,94 +75,39 @@ describe('url.js', () => {
...
-        encodeURI(
-          `${API_PATH}datastore_search?distinct=true&resource_id=${RESOURCE_ID}&fields=${STATICFIELDS}&q=${
-            searchVarsClone.keyword
-          }${categories(searchVarsClone.category)}`,
-        ),
-      );
-    });
-
-    it("doesn't limit when there is no address", () => {
-      searchVarsClone.keyword = keywords.valid;
```
Comment (lamorrison) [inline src/utilities/__tests__/url.test.js]: This behaviour no longer applies. There used to be an increase on the result limit when doing a location search (to ensure more relevant results were fetched from CKAN?)

**Comment by Br3nda** [comment]
This pull request **fixes 1 alert** when merging ff227b4ae19950d7549ab1bcf50b0615639220fb into 534effc3d673fa340eabd5fbb37970133d5dca78 - [view on LGTM.com](https://lgtm.com/projects/g/ServiceInnovationLab/FSD/rev/pr-a0dda4eb70dc43cdc93a5ba1b67b156d57f9dc4a)

**fixed alerts:**

* 1 for Missing variable declaration

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by Br3nda** [comment]
This pull request **fixes 1 alert** when merging e1c4c6c757c5e0fa16d7c29331141566f13f140a into 534effc3d673fa340eabd5fbb37970133d5dca78 - [view on LGTM.com](https://lgtm.com/projects/g/ServiceInnovationLab/FSD/rev/pr-00e89365026ba551816d4dff54457767bcf25f42)

**fixed alerts:**

* 1 for Missing variable declaration

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by Br3nda** [comment]
This pull request **fixes 1 alert** when merging 4e6455f21d6ba0a5370baa1fa3131b5c33061bf7 into 534effc3d673fa340eabd5fbb37970133d5dca78 - [view on LGTM.com](https://lgtm.com/projects/g/ServiceInnovationLab/FSD/rev/pr-13fc7ee1480b88a5ab30e192a8e08bce10fdda6f)

**fixed alerts:**

* 1 for Missing variable declaration

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by Br3nda** [comment]
This pull request **fixes 1 alert** when merging 2e2214fe0463b14c2c88106503af43d6f14c6f65 into 534effc3d673fa340eabd5fbb37970133d5dca78 - [view on LGTM.com](https://lgtm.com/projects/g/ServiceInnovationLab/FSD/rev/pr-bfb60d8aeb7bc3db5b1ac80586aca0398cfd733c)

**fixed alerts:**

* 1 for Missing variable declaration

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by Br3nda** [comment]
Is there no way to tell CKAN to sort by score desc ?

**Comment by Br3nda** [comment]
I've reported this as a bug in CKAN 
https://github.com/ckan/ckan/issues/4660

**Pull request merged**